### ISSUE TITLE
not working via Bluetooth MX keys Mini & MX Anywhere 3

### ISSUE THREAD

**Issue by Varrial** [marcelhoffs/input-switcher#30]
Hi, 
I'm trying to set it up with my mouse and keyboard. 

I got Windows as first device and Macbook as second via Bluetooth.

First I don't know if i'm setting this corectly: 
`.\hidapitester.exe --vidpid 046D:B367 --usage 0x0202 --usagePage 0xff43 --open --length 11 --send-output 0x11,0x00,0x09,0x1E,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00`
in this command I'm changing:
- vidpid 
- usage
- usagePage

with every output line of command `.\hidapitester.exe --list-detail` but it do nothing. 
For example result: `046D/B369:  -
  vendorId:      0x046D
  productId:     0xB369
  usagePage:     0x0001
  usage:         0x0006 [...]`
 I write `.\hidapitester.exe --vidpid 046D:B369 --usage 0x0006 --usagePage 0x0001 --open --length 11 --send-output 0x11,0x00,0x09,[...]`

in this situation I don't know should I change Bytes C and D? I guess yes, but how to get this values on Windows or Mac? 
and should I change anything else if i'm having slighty diffrent mouse and keyboard:
- MX keys Mini 
- MX Anywhere 3

**Comment by purplcha** [comment]
Hi I spent literally 20 hours figuring out how to connect it through bluetooth in my Mac and windows so what I have noticed is that
1. There are multiple usage and usage page but you should always use -- usage 0x0202 --usagePage 0xff43 
2. if you are doing this in a windows then set the length to 11 (--length 11)
3. if you are doing this in a macOS then set the length to 20 (--length 20)
4. for MacOs users, hid can't open keyboards without permission so for it to work add sudo infront of the hidapitester
Hope this helps

**Comment by enieuwy** [comment]
I'm having the same problem on my Macbook.

- MX Keys switches channel 1>2 with: 
`hidapitester --vidpid 046D:B35B --usage 0x0202 --usagePage 0xFF43 --open --length 20 --send-output 0x11,0xFF,0x09,0x1d,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00`

- MX Anywhere 3S doesn't work with
`hidapitester -v --vidpid 046D:B037 --usage 0x0202 --usagePage 0xFF43 --open --length 20 --send-output 0x11,0xFF,0x0a,0x1b,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00`

It writes the output (no error), but doesn't switch.

046D/B037: Logitech - MX Anywhere 3S
  vendorId:      0x046D
  productId:     0xB037
  usagePage:     0x0001
  usage:         0x0002
  serial_number: 64A8F847 
  interface:     -1 
  path: DevSrvsID:4312379593

046D/B037: Logitech - MX Anywhere 3S
  vendorId:      0x046D
  productId:     0xB037
  usagePage:     0x0001
  usage:         0x0001
  serial_number: 64A8F847 
  interface:     -1 
  path: DevSrvsID:4312379593

046D/B037: Logitech - MX Anywhere 3S
  vendorId:      0x046D
  productId:     0xB037
  usagePage:     0xFF43
  usage:         0x0202
  serial_number: 64A8F847 
  interface:     -1 
  path: DevSrvsID:4312379593

**Comment by luiscapobianco** [comment]
This is working for me:

Mouse MX Master 3S
`sudo hidapitester --vidpid ${RCVR_VID}:${MSE_PID} -q --open --length 20 --send-output 0x11,0x00,${MSE_ID},0x1e,${DST_CH},0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00`


Keyboard MX Keys
`sudo hidapitester --vidpid ${RCVR_VID}:${KYB_PID} -q --open --length 20 --send-output 0x11,0x00,${KYB_ID},0x1e,${DST_CH},0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00`

MSE_ID=0x0A
KYB_ID=0x09

DST_CH = destination channel 0x00 for 1, 0x01 for 2, 0x02 for 3

Use hidapitester --list-detail to obtain both the vendorID and productID in my case:
RCVR_VID=046D
KYB_PID=B363
MSE_PID=B034

Ah, without sudo in macOS it fails for one of the devices, I don't remember if it was the mouse or the keyboard. 

I am attaching the script I wrote. 


[switch.txt](https://github.com/user-attachments/files/17970695/switch.txt)

**Comment by enieuwy** [comment]
Great script, thank you!
Unfortunately it still isn't working for my mouse. I suspect my MSE_ID may be different. Where do I find this (for bluetooth, not unifying software)?

**Comment by luiscapobianco** [comment]
Did you use sudo to run the command?

**Comment by enieuwy** [comment]
Yes, wouldn't work without sudo. Added the program to the sudoers list.
Also managed to package everything in a Mac Shortcut (including monitor input switch) - no script on the Windows machine! It's run via ssh.
<img width="618" alt="InputSwitch" src="https://github.com/user-attachments/assets/7c46ea7e-7bf3-4d19-b3fc-5b69178d9cae">

**Comment by purplcha** [comment]
This is how I got mine working on Mac
1. I enabled sudo without password by following this link :https://jefftriplett.com/2022/enable-sudo-without-a-password-on-macos/
This is required if you want to run the script without typing the password all the time because MacOS need permission for keyboard because it can cause privacy concerns.

2. This is my code after I did step 1 
```
#!/bin/bash 

# Switch MX Keys to other device
sudo hidapitester --vidpid 046D:B35B --usage 0x0202 --usagePage 0xff43 --open --length 20 --send-output 0x11,0x00,0x09,0x1E,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00
# Switch MX Master 3s to other device                                                                             
hidapitester --vidpid 046D:B034 --usage 0x0202 --usagePage 0xff43 --open --length 20 --send-output 0x11,0x00,0x0A,0x1E,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00

exit 0
```
After this I did everything the readme file mentioned in the Mac folder, like auto closing the terminal when the script runs and all that stuff
